crates/toml/src/de.rs

```diff
@@ -96,6 +142,21 @@ struct ErrorInner {
     key: Vec<String>,
 }
 
+/// Error that can occur when deserializing via IO.
+#[derive(Debug)]
+pub enum ReaderError{
```

Comment:
How come this is using a custom error type?

Reply:
Please come back with a proposal on that.  I would prefer to keep to a single error type especially because others, at least serde_json,. get away with one error type.

Reply:
Using the same error for both doesn't really seem to make sense TBH. For example:

- When serialising, there's no `line` and `column` related to the error, but this must be allocated and returned in the error anyway.
- The [`Error.category`](https://docs.rs/serde_json/latest/serde_json/error/enum.Category.html) has categories that don't apply to both operations, but application code using the library needs to handle both. When serialising, `Category::Syntax` does not make sense as a possible return value, but needs to be handled anyway. Of course, it's possible for usages to simply use `Category::Syntax => unreachable!()`, but that's an indicator of a poor API.

Reply:
I guess we could use `line: Option<usize>` and use more generic categories. Like `Category::InvalidData` when either the `JSON` input is invalid, or when the input data cannot be serialised into json.